### ISSUE TITLE
wat

### ISSUE THREAD

**Issue by SylvainCorlay** [numpy/numpy#7967]
Here is a list of surprising behaviors of numpy. There are certainly some deep reasons for most of them.

1. Initialize a 0-D array with None

    ```python
    >>> np.array(1) ==  1
    True
    >>> np.array(None) ==  None
    False  # wat?
    ```

    Regardless on whether comparing with None is good or not, this is surprising to the user that the behaviors differ.

    I came across this where the value passed were in variables...

2. Ufuncs return types for 0-D arrays

    ```python
    >>> np.exp(1).__class__
    numpy.float64   # ok: scalar input scalar output

    >>> np.exp(np.array([1])).__class__
    numpy.ndarray  # ok: array input, array output

    >>> np.exp(np.array(1)).__class__
    numpy.float64  # wat? 0-D array input, scalar output
    ````
   
    I think that the return type of ufuncs should be arrays in case of array-like objects.

3. ndarray constructor with a shape

    ```
    >>> np.ndarray(shape=(0, 0))
    array([], shape=(0, 0), dtype=float64)  # ok

    >>> np.ndarray(shape=(1, 1))
    array([[0.]])  # ok 

    >>> np.ndarray(shape=(1, 0))
    array([], shape=(1, 0), dtype=float64)  # wat? 
    ```

    why is the repr no equal to `array([[]], shape=(1, 0), dtype=float64)` in this case?

    Similarly
    ````
    >>> bool(np.ndarray(shape=(0, 0)))
    False  # ok

    >>> np.ndarray(shape=(1, 0))
    Flase # wat?
    ```
    why is the latter a falsy value? it has a shape of `(1, 0)`. Note: I do expect an array of shape `(0, 1)` to be falsy because an empty list *is*.

4. ndarray default constructor

    This one is more of a general question: I generally expect value-semantic objects to have an argument-less default constructor returning a falsy value. 

    Could np.ndarray() return a 0-D array holding the 0.0 value?

**Comment by shoyer** [comment]
For better or for worse, NumPy intentionally violates Python's norms for implicit boolean values (among other things). It supports Boolean casting on scalars, but anything else (including these edge cases with size 0 arrays) should raise TypeError. So 4 and 5 are probably "won't fix".

1 and 3 look like bugs. 2 and 6 also look like bugs, but apparently reflect a long standing issue where many NumPy functions don't distinguishing between arrays and numpy scalars. Arguably it was a mistake to make the separate scalar class in the first place, but it's way too late to change that for numpy now. In any case, it would be great if you could dive into any of these.

**Comment by seberg** [comment]
1. Not quite sure what you mean, the warnings means that it will return True in the future?

2. Agree with @shoyer a long standing issue and something a little tricky, I fully agree that numpy should not convert 0d arrays to scalars implicitly, but I am not sure it is something one can fix easily.

3. The representation could be nicer here, but I think you should not think of it as a list of lists anyway.  Which also leads to the False making sense, since the array is a *single* container, and an empty container is considered False. The problem is that numpy uses a "list of lists" like behaviour for default iteration, so that we have the notion of a nested container in some areas....

4. We could do that I suppose, though I am not sure what a good default would be, an empty array might even make more sense.

5. I think this would be controversial and is general discussion. I can see that for many cases it is inconvenient, but lacking another operator and noting that in most code the `==` logic is far more common, I doubt you can convince a majority of users that it is better (and even if, I don't think changing it will be an option). Which leads me to:
   * What *other* option could improve this? Maybe we have to "work around" numpy, but what if we introduce an n-d container ABC, and things serializers could figure out what to do by numpy registering with it?

6. Not exactly sure, I think its probably because `np.int64` is a subclass of `int` here, plus the annoying 0-d array to scalar logic. I wouldn't mind discussing changing that logic in this case more specifically or even more generally, but it might be a rather large discussion.

@shoyer just a note, I am not really convinced that scalars are a bad idea. Duplicating so much implementation, probably. Converting to scalars implicitly, definitely. But as such, scalars are for example immutable, etc.

**Comment by seberg** [comment]
What I am saying is that this sounds like some of these things are frustrating, and while I don't think we can change all of them (especially soon), I am sure there are other solutions/approaches available.

**Comment by SylvainCorlay** [comment]
> Not quite sure what you mean, the warnings means that it will return True in the future?

I meant two things (sorry if it was not clear):
 1. since `array(None)` is a 0-D array, element-wise comparison is the same as equality, and therefore, there should not be a warning when comparing a 0-D array to None.
 2. The correct return value is True.

> Agree with @shoyer a long standing issue and something a little tricky, I fully agree that numpy should not convert 0d arrays to scalars implicitly, but I am not sure it is something one can fix easily.

Yes, even though there are performance benefits to using the immutable scalar type as a return value instead of a 0-D array, I think that it is terrible that a single function can return objects with so different semantics depending on the value of the parameters. Besides, at a time where JITs are becoming more mainstream, this special casing is probably going to be an issue.

**Comment by seberg** [comment]
Ah well, the thing is that `array(None)` thing is as much fixed as possible at this time, since that is what the FutureWarning is for.

I personally think even if you ignore performance things (not sure if it really matters), there are places where it makes sense and some where it does not. The typical way to create a new scalar is something like `arr.sum()`, where not preserving the array makes perfect sense, e.g. `arr.sum((0, 1))` could return an array, while `arr.sum(None)` could return a scalar (assuming arr is two dimensional). But making an existing 0-d array a scalar on an elementwise operation without asking does not make sense.
Immutability itself is not just performance issues, it is also about later in-place manipulation of the original array and hashibility.

**Comment by Darkdragon84** [comment]
I am totally in favor of addressing 5! It has been THE major PITA for me when using `np.ndarray`s. Any other sensible class implements `__eq__`  as returning `bool` and (usually) non-raising. 

In my understanding, `==` should mainly check if the underlying data of the compared objects are equal. all of it. If you want element-wise, you should explicitly call `numpy.equal` imho. 

Because of numpy's weird behavior I always have to do very nasty and ugly workarounds with `if isinstance(x, np.ndarray)`. For example, when used in `dataclasses, a separate `__eq__` needs to be implemented and subclasses of that dataclass need to specifically be decorated with `dataclass(eq=False)`. That alone unnecessarily removes a lot of the convenience of dataclasses. 
There are plenty of other examples and I am sure everyone who ever worked seriously with `np.ndarray`s can tell similar stories.

I understand pointing `==` to `np.array_equal` instead of `np.equal` would break a lot of existing user code. But I dare to assume this will be mostly ugly workaround code anyway. Please consider making everyone's lives easier by making numpy consistent with usual python behavior at least in this single instance and have `==` point to `np.array_equal` :pray:

**Comment by rkern** [comment]
Sorry, the current behavior of `==` is definitely _widely_ used as designed and not just ugly workaround code. Changing it will break tons of code.

**Comment by shoyer** [comment]
I agree, it's way too late to change == for NumPy.

**Comment by SylvainCorlay** [comment]
I agree too. I listed it here originally as it is somewhat related to the
other items listed.

**Comment by Darkdragon84** [comment]
I agree that the assumption that most broken code will be workaround code might have been a strong one, that's why I said I dare postulate. I just speak from my and all my colleagues' experience. No insult intended and sorry if it was understood that way :hugs:. And let me be clear that otherwise I love numpy. It's a fantastic library and everyone contributing to it is doing a great job.

I would like to respectfully disagree however, that it's too late to change `==`'s behavior. Imho it is never too late to improve a library and make it more convenient to use for everyone. There are many breaking changes that have been introduced before and that have been advertised with deprecation and future warnings ahead of time. I don't see why this should be so much different.

Let me make this clear. The pain is real. I don't know a single person that wouldn't prefer `==` to point to `np.array_equal`.

Let me argument another way:
So, the element-wise behavior of `==` is numpy specific. Any code that assumes that behavior works specifically with and on `np.ndarray`s. I am not aware of any other library with classes that implement a similar interface. So, if the behavior of `==` were to change it would be a very simple refactoring task for any existing code to just replace any `arr1 == arr2` with `equal(arr1, arr2)` for `np.array`s only.
On the other hand, the behavior of `==` to return a single bool is the widely used standard everywhere in Python. Especially in applications that work with many different data types (among which `np.array`s), that might not be known at implementation time, it would be a huge relief to align numpy with standard Python behavior.
So, in short, a change of `==` would entail one refactoring step of existing code that explicitly works with `np.array`s, but would just simplify all mixed data type applications a ton. In my opinion the sacrifice of the former is definitely worth the benefit of the latter. And it would certainly be more pythonic imho.

If you disagree I would be very interested to read a more detailed elaboration as to why you think so. And especially I'd love some pointers (besides the `if isinstance(obj, np.ndarray)`) how to make the lives of people easier who have to navigate around the element-wise behavior of `==` on a regular basis (subclassing, ...). Happy with SO links too :smiley: Thanks a bunch!

**Comment by rkern** [comment]
> No insult intended and sorry if it was understood that way .

None was taken! I just wanted to clarify the prevalence of this feature in my long experience writing and reading numpy code. Most non-trivial numpy programs I write, for instance, use this somewhere in them. The entire reason that Python allows `__eq__` and friends to return something other than a `bool` is because we lobbied for it. Being able to build these expressions using real comparison operators rather than function calls was a significant improvement in the readability of these expressions.

There are other array-like objects (like PyTorch and Tensorflow `Tensor`s) that follow our example (no surprise, as they are trying to mimic our API in large part). They would need their own replacements, not necessarily `np.equal()` (the recent work on ufunc dispatching might make that _possible_, though). Other non-numeric libraries do use rich comparisons to return non-boolean results, like [SQLAlchemy](https://docs.sqlalchemy.org/en/14/core/operators.html).

I do recognize the annoyance that it can cause in some situations (mostly checking for containment in lists, in my experience, as there's little way to work around it since the comparison is performed inside `list.__contains__`). For me, at least, those are rarer than doing the element-wise comparisons, and the tradeoff is worth it to me. Also, when it does show up, I find that it's a little ambiguous as to what I actually want to compare equal in those cases. Sometimes, I want broadcasting to work, sometimes not. Sometimes, I want cross-type comparisons to work (`3 == 3.0`), sometimes not. Sometimes, I wish I could just set `list.__contains__` to use object identity instead of equality comparisons. In some ways, these cases are like trying to rely on `__hash__`: some objects implement it and some don't and the places that _use_ it don't give me much hooks to customize things to work the way I want them to.

I don't quite know what kind of "mixed data type applications" you are working with. There's the case where you are just trying to write code that works with any object (storing it in a cache, keeping track of objects that you've seen in a list, etc.). As I mentioned above regarding `list.__contains__`, that can particularly be a source of annoyance. If you are explicitly writing out the `==`, a workaround is available in that you can define an `eq()` function that checks for `ndarray` (and maybe some of the others) and uses whatever's appropriate for them. Then your actual code won't be littered with such checks.

**Comment by matthew-brett** [comment]
@Darkdragon84 - Numpy has the same behavior in this respect as R and Matlab:

R:

```
> c(1, 2, 3) == c(1, 0, 3)
[1]  TRUE FALSE  TRUE
```


Matlab (and Octave):

```
>> [1 2 3] == [1 0 3]

ans =

     1     0     1
```

**Comment by seberg** [comment]
The elementwise operation of NumPy arrays for comparisons probably has to be assumed an unshakable given.  It is far too convenient and heavily used in the community.

I expect that if we did not have it, we would get more request the opposite direction honestly.  Equality is special, but I don't understand why it should be special enough to avoid the generic pattern.  It is a very nice pattern in "array programming" (I personally prefer the word elementwise.)

> And especially I'd love some pointers (besides the if isinstance(obj, np.ndarray)) 

The only thing I see right now that might help would be to try to push a small change, probably in Python itself.  That could be `isinstance(obj, ElementwiseOperationsContainer)` as an ABC.  Or having an "equivalent" operation (something that has been rejected in Python before – mostly discussed in the context of NaNs to formalize/fix the `NaN in list` logic).

Basically, I am just throwing things out there.  I don't know how it should look like, because I have poor understanding of what you actually struggle with here.
(E.g. I think having an `ElementwiseOperationsContainer` to quickly recognize torch, etc. tensors might be neat in rare cases, but overall probably doesn't do much for NumPy itself.)

The main issues that I feel I remember are from library authors implementing serialization.  It seems you have a different reason, but it is not clear to me what that reason is.  Maybe writing a non-element-wise container using the NumPy element-wise container as a basis?

**Comment by Darkdragon84** [comment]
Hey @seberg @rkern and @matthew-brett  thanks for your considerate answers! Especially @rkern thanks for the background elaboration, it makes more sense to me now. A bit of background on my current pain, which comes from serialization using dataclasses and tryin to accommodate for all possible field datatypes (so, not a particular dataclass with known fields and types). Dataclasses conveniently autogenerate their own `__eq__` using each field's own `__eq__`. So with numpy arrays somewhere in there, possibly nested in lists, tuples and dicts or other containers, you have to iterate through all the leaves of that object tree and disambiguate at the leaf level, circumventing also the containers' `__eq__` methods. But we are considering attrs instead now, which apparently lets you define a custom comparison method for each field, so that could solve that issue. Yeah it seems to me (and it has been mentioned above) that Python itself could use another operator for equivalence (I always interpret `==` to be just that)

>  Equality is special, but I don't understand why it should be special enough to avoid the generic pattern
I think my main point is that since any Python class automatically subclasses `object`, whose `__eq__` signature clearly returns a single `bool`, any custom class in Python should stick to that signature if in any way possible (Yes @matthew-brett I know that R and Matlab also follow that paradigm, but that's what these languages are based upon, while Python is generally not.)
Now I understand the choice not to do so was made a long time ago (and @rkern motivated that choice, which makes total sense to me, thanks!) and there is just too much code now that clearly benefits from the current implementation of `__eq__`. So yes, let's have another operator for equivalence please :grin:
